storage: don't spuriously retry lease proposals

tbg opened the pull request:
When the Replica notices that an in-flight proposal has a
`MaxLeaseIndex` not strictly ahead of the `LeaseAppliedIndex`, the proposal
is considered failed (since it won't be able to apply successfully).

The one exception are lease requests, for which `MaxLeaseIndex` is
irrelevant. We weren't taking into account that exception, which resulted
in the following odd sequence of events observed when a range was woken up
from a dormant state:

1. lease gets proposed (at some bogus low `MaxLeaseIndex`) 2. Raft group
goes through elections 3. refreshProposalsLocked deletes the proposal
because it looked failed 4. lease gets proposed again 5. at apply time,
previous lease request has already updated the lease,
so new proposal fails.

In effect, this caused an extra round of consensus.

The test added in this commit prevents regression of this problem. It fails
without the change in `refreshProposalsLocked`.

Release note: None

tbg committed:
storage: test lease request MaxRaftIndex handling

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26509)
<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This gets me wondering why we assign a `MaxLeaseIndex` to lease commands at all. 

---

Reviewed 1 of 1 files at r1.
Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/storage/replica.go, line 4233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26509#-LEPrSxO7cPztzzmtl2N:-LEPrSxP74MR8UlajEb4:b-jgjt55) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee18fce00897272dab649ab1841362e8c8b7b7be/pkg/storage/replica.go#L4233)):*
> ```Go
> 					fmt.Sprintf("unknown status for command without MaxLeaseIndex "+
> 						"at refreshProposalsLocked time (refresh reason: %s)", reason)))})
> 		} else if !p.command.ReplicatedEvalResult.IsLeaseRequest && p.command.MaxLeaseIndex <= r.mu.state.LeaseAppliedIndex {
> ```

Add somewhere at the end of this long comment why we have this second condition.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26509#-:-LEPrgef2aVSxaeYQQRV:b-4d00ei)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I saw the same phenomenon and wrote it up in #26448. I was thinking the fix would be to cancel pending requests when a new lease is applied instead of skipping the refresh here. Do we still need something like that or can we consider #26448 closed by this PR?

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26509#-:-LEPriCA3UsCoCd7kQx2:b-9gx73z)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't spuriously retry lease proposals

When the Replica notices that an in-flight proposal has a
`MaxLeaseIndex` not strictly ahead of the `LeaseAppliedIndex`, the
proposal is considered failed (since it won't be able to apply
successfully).

The one exception are lease requests, for which `MaxLeaseIndex` is
irrelevant. We weren't taking into account that exception, which
resulted in the following odd sequence of events observed when a
range was woken up from a dormant state:

1. lease gets proposed (at some bogus low `MaxLeaseIndex`)
2. Raft group goes through elections
3. refreshProposalsLocked deletes the proposal because it looked failed
4. lease gets proposed again
5. at apply time, previous lease request has already updated the lease,
   so new proposal fails.

In effect, this caused an extra round of consensus.

The test added in this commit prevents regression of this problem.
It fails without the change in `refreshProposalsLocked`.

Release note: None

tschottdorf commented:
@bdarnell there are no competing lease requests here so I'm not sure. I commented on that issue to understand the situation there better.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 4233 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26509#-LEPrSxO7cPztzzmtl2N:-LEW2d2_DmE9JEcXkL7o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee18fce00897272dab649ab1841362e8c8b7b7be/pkg/storage/replica.go#L4233)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add somewhere at the end of this long comment why we have this second condition.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26509#-:-LEW2Y0Z54Xm5YvcXeD9:blej6it)*
<!-- Sent from Reviewable.io -->
